File: crates/viben-core/src/gateway/state.rs
//! Application state

use std::sync::Arc;

use crate::channels::{ChannelRouter, ChannelService};
use crate::db::DbService;
use crate::services::{
    ContainerService, CronService, EventService, HistoryService, PtyService, SessionStoreService,
};

/// Application state shared across handlers
#[derive(Clone)]
pub struct AppState {
    /// Database service
    pub db: Arc<DbService>,
    /// Event service for SSE
    pub events: Arc<EventService>,
    /// Container service for process management
    pub container: Arc<ContainerService>,
    /// PTY service for terminal emulation
    pub pty: Arc<PtyService>,
    /// History service for .agent_history management
    pub history: Arc<HistoryService>,
    /// Session store service for file-based session persistence
    pub session_store: Arc<SessionStoreService>,
    /// Cron service for scheduled task management
    pub cron: Arc<CronService>,
    /// Channel service for channel instance management
    pub channel: Arc<ChannelService>,
    /// Channel message router
    pub channel_router: Arc<ChannelRouter>,
}

impl AppState {
    /// Create a new application state
    pub fn new(
        db: DbService,
        events: EventService,
        container: ContainerService,
        pty: PtyService,
        history: HistoryService,
        session_store: SessionStoreService,
        cron: CronService,
        channel: ChannelService,
        channel_router: ChannelRouter,
    ) -> Self {
        let events = Arc::new(events);
        Self {
            db: Arc::new(db),
            events: events.clone(),
            container: Arc::new(container),
            pty: Arc::new(pty),
            history: Arc::new(history),
            session_store: Arc::new(session_store),
            cron: Arc::new(cron),
            channel: Arc::new(channel),
            channel_router: Arc::new(channel_router),
        }
    }

    /// Create application state with defaults
    pub async fn with_defaults() -> Result<Self, crate::db::DbError> {
        tracing::debug!(target: "viben::gateway::state", "Initializing DbService...");
        let db = DbService::new().await?;
        tracing::debug!(target: "viben::gateway::state", "DbService initialized");

        tracing::debug!(target: "viben::gateway::state", "Initializing EventService...");
        let events = EventService::new();
        let events_arc = Arc::new(events);
        tracing::debug!(target: "viben::gateway::state", "EventService initialized");

        tracing::debug!(target: "viben::gateway::state", "Initializing ContainerService...");
        let container = ContainerService::new((*events_arc).clone());
        let container_arc = Arc::new(container);
        tracing::debug!(target: "viben::gateway::state", "ContainerService initialized");

        tracing::debug!(target: "viben::gateway::state", "Initializing PtyService...");
        let pty = PtyService::new();
        tracing::debug!(target: "viben::gateway::state", "PtyService initialized");

        tracing::debug!(target: "viben::gateway::state", "Initializing HistoryService...");
        let history = HistoryService::new();
        tracing::debug!(target: "viben::gateway::state", "HistoryService initialized");

        tracing::debug!(target: "viben::gateway::state", "Initializing SessionStoreService...");
        let session_store = SessionStoreService::new();
        tracing::debug!(target: "viben::gateway::state", "SessionStoreService initialized");

        tracing::debug!(target: "viben::gateway::state", "Initializing CronService...");
        let cron = CronService::new(events_arc.clone());
        tracing::debug!(target: "viben::gateway::state", "CronService initialized");

        tracing::debug!(target: "viben::gateway::state", "Initializing ChannelService...");
        let channel = ChannelService::new(events_arc.clone());
        // Load channels from config file
        if let Err(e) = channel.load().await {
            tracing::warn!(
                target: "viben::gateway::state",
                "Failed to load channel config: {}",
                e
            );
        }
        let channel_arc = Arc::new(channel);
        tracing::debug!(target: "viben::gateway::state", "ChannelService initialized");

        tracing::debug!(target: "viben::gateway::state", "Initializing ChannelRouter with ContainerService...");
        let channel_router = ChannelRouter::with_container(
            events_arc.clone(),
            channel_arc.clone(),
            container_arc.clone(),
        );
        tracing::debug!(target: "viben::gateway::state", "ChannelRouter initialized");

        tracing::info!(
            target: "viben::gateway::state",
            "All services initialized: db, events, container, pty, history, session_store, cron, channel, channel_router"
        );

        Ok(Self {
            db: Arc::new(db),
            events: events_arc,
            container: container_arc,
            pty: Arc::new(pty),
            history: Arc::new(history),
            session_store: Arc::new(session_store),
            cron: Arc::new(cron),
            channel: channel_arc,
            channel_router: Arc::new(channel_router),
        })
    }
}

File: packages/core/src/gateway/state.ts
/**
 * Gateway application state
 *
 * Holds all the services and shared state for the gateway server.
 */
import { EventService } from "../services/events";
import { SessionStoreService } from "../services/session-store";
import { CronService } from "../cron";
import { ContainerService } from "../services/container";
import { HistoryService } from "../services/history";
import { MessageBus } from "../services/message-bus";
import { ChannelRouter, ChannelRuntime, channelManager } from "../channels";
import { TaskQueueManager } from "./queue";
import { TaskRecoveryService } from "../task/recovery/task-recovery";
import { taskEventStore } from "../task/events/event-store";
import { TaskSSEManager } from "./sse/task-sse-manager";
import { CommandQueue } from "../queue/core/command-queue";
import { DeviceRegistryService } from "../devices/device-registry";
import { MeshService } from "../mesh/mesh-service";
import { PeerStore } from "../mesh/peer-store";
import { DiscoveryService } from "../discovery/discovery-service";
import { ClientStore } from "./client-store";
import type { ClientSocketServer } from "./client-socket-server";
import { FirebaseService } from "../services/firebase";

/**
 * Application state for the gateway
 */
export interface AppState {
  /** Event service for SSE/WebSocket streaming */
  events: EventService;
  /** Session store service for file-based session persistence */
  sessionStore: SessionStoreService;
  /** Cron service for scheduled jobs */
  cron: CronService;
  /** Container service for process management */
  container: ContainerService;
  /** History service for agent history */
  history: HistoryService;
  /** Message bus for channel routing */
  messageBus: MessageBus;
  /** Channel router for routing messages to bound agents */
  channelRouter: ChannelRouter;
  /** Channel runtime for managing polling clients */
  channelRuntime: ChannelRuntime;
  /** Task queue for concurrent agent execution control */
  taskQueue: TaskQueueManager;
  /** Task recovery service for stuck task detection */
  taskRecovery: TaskRecoveryService;
  /** Task SSE manager for real-time task state updates */
  taskSSEManager: TaskSSEManager;
  /** Command queue for detached shell command execution (viben task start, etc.) */
  commandQueue: CommandQueue;
  /** Device registry for tracking all mesh devices */
  deviceRegistry: DeviceRegistryService;
  /** Mesh service for gateway-to-gateway connections */
  mesh: MeshService;
  /** Discovery service for mDNS and QR code */
  discovery: DiscoveryService;
  /** Client store for Socket.io connected clients and their actions */
  clientStore: ClientStore;
  /** Client Socket.io server (set after onReady) */
  clientSocketServer?: ClientSocketServer;
  /** Firebase service for push notifications, bug reporting, and analytics */
  firebase: FirebaseService;
}

export interface AppStateConfig {
  /** Gateway host (default: "127.0.0.1") */
  host?: string;
  /** Gateway port (default: 18790) */
  port?: number;
  /** Start background runtime services */
  runtime?: boolean;
}

/**
 * Create application state with default services
 */
export function createAppState(config: AppStateConfig = {}): AppState {
  const { host = "127.0.0.1", port = 18790, runtime = true } = config;
  const events = new EventService();
  const sessionStore = new SessionStoreService();
  const cron = new CronService(events);
  const container = new ContainerService(events, sessionStore);
  const history = new HistoryService();
  const messageBus = new MessageBus(events);

  // Create channel router with container service for agent execution
  const channelRouter = new ChannelRouter({
    events,
    channels: channelManager,
    container,
    responseTimeout: 120000, // 2 minutes timeout for agent responses
  });

  // Create channel runtime for polling (receives messages from external channels)
  const channelRuntime = new ChannelRuntime({
    channelManager,
    messageBus,
    auto_start: true, // Auto-start enabled channels with agent bindings
    pollingTimeout: 30,
  });

  // Create task queue manager
  const taskQueue = new TaskQueueManager(events);

  // Create task SSE manager for real-time task state updates
  // Configure with automatic stale subscriber cleanup
  const taskSSEManager = new TaskSSEManager({
    heartbeatIntervalMs: 30000, // 30 seconds heartbeat
    staleTimeoutMs: 120000, // 2 minutes stale timeout
    maxFailedSends: 3, // 3 failed sends = dead connection
    cleanupIntervalMs: 60000, // 1 minute cleanup interval
  });
  if (runtime) {
    taskSSEManager.startHeartbeat();
  }

  // Create task recovery service for stuck task detection
  const taskRecovery = new TaskRecoveryService(taskEventStore, taskSSEManager, {
    stuckThresholdMs: 5 * 60 * 1000, // 5 minutes
    autoRecover: true,
  });

  // Create command queue for detached shell command execution
  // This manages "viben task start <task>" commands in the background
  const commandQueue = new CommandQueue();

  // Create device registry
  const deviceRegistry = new DeviceRegistryService(events);
  const gatewayId = deviceRegistry.getGatewayId();

  // Create peer store for YAML persistence
  const peerStore = new PeerStore();

  // Create mesh service
  const localInfo = {
    gateway_id: gatewayId,
    name: `viben-${gatewayId.slice(0, 8)}`,
    version: "1.0.0",
    capabilities: ["navigate", "notify", "ping"],
    address: `http://${host}:${port}`,
  };
  const mesh = new MeshService(events, deviceRegistry, peerStore, localInfo);

  // Create client store for Socket.io clients
  const clientStore = new ClientStore();

  // Create discovery service
  const discovery = new DiscoveryService(events, {
    gateway_id: gatewayId,
    name: localInfo.name,
    version: "1.0.0",
    port,
  });

  // Create Firebase service (gracefully degrades if not configured)
  const firebase = new FirebaseService();

  if (runtime) {
    discovery.onPeerDiscovered((address) => {
      mesh.connectToPeer(address);
    });
  }

  return {
    events,
    sessionStore,
    cron,
    container,
    history,
    messageBus,
    channelRouter,
    channelRuntime,
    taskQueue,
    taskRecovery,
    taskSSEManager,
    commandQueue,
    deviceRegistry,
    mesh,
    discovery,
    clientStore,
    firebase,
  };
}
